[Loh Wei-Ming, Darin] ip

talkintomato opened the pull request:
# Duke is the task manager of task managers 
> The task manager your task manager uses :sunglasses:

## Reasons to use Duke: 
1. It will _never_ give you up 
2. It will _never_ let you down 

## Things Duke cannot do: 
- Feed your cat 
- [Look at all those chickens ](https://www.youtube.com/watch?v=NsLKQTh-Bqo&ab_channel=rawuncutvines)

## Features: 
- [x] create tasks 
- [ ] set deadlines 
- [ ] be your friend 

``` java
public class Duke {
    Run();
}
```

Commit:
Add Gradle support

talkintomato committed:
First commit Week 2 Level-1 Complete

talkintomato committed:
Level-2 added listing function

talkintomato committed:
Added mark and unmark functions

talkintomato committed:
Level-4 done

talkintomato committed:
Level-4 revised

talkintomato committed:
level-4 v3

talkintomato committed:
basic functionality complete need to finish cleaning up edge cases

talkintomato committed:
Week 2 complete

talkintomato committed:
testing complete

talkintomato committed:
Level 7 Save

talkintomato committed:
added minimal Level 8

talkintomato committed:
rename branch

talkintomato committed:
resolved merge conflict

talkintomato committed:
A-MoreOOP

talkintomato committed:
A-packages

talkintomato committed:
Test written for TaskList and EventTest

talkintomato committed:
JAR file created

talkintomato committed:
JavaDocs added for some methods

talkintomato committed:
Find methods completed

talkintomato committed:
Minor edits

talkintomato committed:
resolved A-javaDoc branch merge

talkintomato committed:
resolved conflicts with level-9 branch

talkintomato committed:
minor changes

talkintomato committed:
Merge branch 'add-gradle-support' into master

KeithCZW commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,130 @@
+
+    private String[] cmd;
+
+    private boolean isExit = false;
```
I feel that isExit is not descriptive enough. Is the command object an exit of what? Perhaps isExiting would be a better name for this.

KeithCZW commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,130 @@
+     * @throws DukeException
+     */
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        switch(cmd[0]) {
+            case "list" :
+                ui.printList(taskList.getList());
+                break;
+            case "mark" :
+                taskList.mark(cmd[1]);
+                break;
+            case "unmark" :
+                taskList.unmark(cmd[1]);
+                break;
+            case "find" :
+                ui.showSearchResults(taskList.searchKeyword(cmd[1]));
+                break;
+            //Event creation
+            case "todo":
+                createTodo(cmd, taskList, ui);
+                break;
+            case "deadline":
+                createDeadline(cmd, taskList, ui);
+                break;
+            case "event":
+                createEvent(cmd, taskList, ui);
+                break;
+            case "delete":
+                taskList.removeTask(cmd[1]);
+                break;
+            case "bye" :
+                isExit = true;
+                break;
+            default:
+                throw new DukeException("I'm not sure what that means :((");
```
The indentations are off for your switch cases. The case should be inline with the switch function.

talkintomato replied:
@KeithCZW Thanks for the feedback! do u have a link which shows the case and switch being inline? most of the examples I have googled have shown otherwise.

KeithCZW replied:
https://se-education.org/guides/conventions/java/intermediate.html#layout

Under layout section, switch and cases are aligned.

talkintomato replied:
Thanks!

KeithCZW commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,14 @@
+
+public class Parser {
+
+
```
Just a small nitpick, but there's an additional space here. 😅

KeithCZW commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,80 @@
+public class Storage {
+
+    protected ArrayList<Task> todos = new ArrayList<>();
+    private String filepath;
```
filepath should be filePath to align with camel case.

KeithCZW commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,80 @@
+                }
+                else if (currTask instanceof Deadline) {
+                    Deadline currDeadline = (Deadline) currTask;
+                    myWriter.write(currDeadline.getInitial() + " , " + currDeadline.getIsDone() + " , " + currDeadline.getDescription() + " , " + currDeadline.getTime() +"\n");
```
This line is more than 120 characters and should be split into two lines.

KeithCZW reviewed (commented):
Just a few minor issues with the PR but overall it's looking good! Just a few small parts to make changes to in order to follow the coding standard. Good job!

talkintomato committed:
minor styling changes made

talkintomato committed:
first commit

jaysmyname commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,130 @@
+package duke;
+import duke.exception.DukeException;
+import duke.tasks.*;
```
Shouldn't imported classes be listed explicitly?

jaysmyname commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,130 @@
+        storage.saveList();
+    }
+
+    public boolean getExitStatus() {
```
Perhaps change the boolean method to sound more like a boolean? Something like shouldExit()

jaysmyname commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,83 @@
+                Task currTask = todos.get(i);
+                if (currTask instanceof Todo) {
+                    Todo currTodo = (Todo) currTask;
+                    myWriter.write(currTodo.getInitial() + " , " + currTodo.getIsDone() + " , "
+                            + currTodo.getDescription() + " , " + currTodo.getTime() +"\n");
+                }
+                else if (currTask instanceof Event) {
+                    Event currEvent = (Event) currTask;
+                    myWriter.write(currEvent.getInitial() + " , " + currEvent.getIsDone() + " , "
+                            + currEvent.getDescription() + " , " + currEvent.getTime() +"\n");
+                }
+                else if (currTask instanceof Deadline) {
+                    Deadline currDeadline = (Deadline) currTask;
+                    myWriter.write(currDeadline.getInitial() + " , " + currDeadline.getIsDone() + " , "
+                            + currDeadline.getDescription() + " , " + currDeadline.getTime() +"\n");
```
Perhaps add a method to each task type to return a string that has all the gets so you can simply do one get call?

jaysmyname commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,130 @@
+     * @throws DukeException
+     */
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
+        switch(cmd[0]) {
+        case "list" :
+            ui.printList(taskList.getList());
+            break;
+        case "mark" :
+            taskList.mark(cmd[1]);
+            break;
+        case "unmark" :
+            taskList.unmark(cmd[1]);
+            break;
+        case "find" :
+            ui.showSearchResults(taskList.searchKeyword(cmd[1]));
+            break;
+        //Event creation
+        case "todo":
+            createTodo(cmd, taskList, ui);
+            break;
+        case "deadline":
+            createDeadline(cmd, taskList, ui);
+            break;
+        case "event":
+            createEvent(cmd, taskList, ui);
+            break;
+        case "delete":
+            taskList.removeTask(cmd[1]);
+            break;
+        case "bye" :
+            isExiting = true;
+            break;
+        default:
+            throw new DukeException("I'm not sure what that means :((");
+        }
```
I like how concise this is! Very compact.

jaysmyname commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.Scanner;
+
+import duke.exception.DukeException;
+import duke.tasks.*;
```
Shouldn't imported classes be listed explicitly?

jaysmyname reviewed (commented):
Overall really good job following the coding standards!

talkintomato committed:
GUI has been added

talkintomato committed:
command is now an abstract parent class

each command is its own object which inherits from the abstract class
command as the commands have similar methods.

let's use parser to translate text input from user into an object which
inherits from command. using polymorphism when deadling with commands
allows for neater and more readable code.

Using inheritance is preferable over composition in this situation
becasuse behaviours are not composable.

Oddcorner commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,148 @@
+        return this.isExiting;
+    }
+
+//    public void validateTaskCreation(String[] cmd) {
+//        if (cmd.length != 2) {
+//    }
```
Perhaps it would be better to remove this code entirely instead of leaving it commented. I noticed a few other places where quite a large number of unused code lines are commented too.

Oddcorner commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,148 @@
+     * @throws DukeException
+     */
+    private String createDeadline(String[] cmd, TaskList taskList) throws DukeException {
+        String response = "";
+        if (cmd.length == 2) {
+            String[] deadlineDetails = cmd[1].split(" /by ", 2);
+            if (deadlineDetails.length == 2) {
+                try {
+                    Deadline deadline = new Deadline(deadlineDetails[0], deadlineDetails[1]);
+                    taskList.addTask(deadline);
+                    response = ui.showTaskAdded(deadline);
+                } catch (DateTimeParseException e) {
+                    System.out.println(e.getMessage() + "\nPlease input deadline in yyyy-mm-dd");
+                }
+            } else {
+                throw new DukeException("Please indicate a time for the Deadline!");
+            }
+        } else {
+            throw new DukeException("The description of a Deadline date cannot be empty.");
+        }
+        return response;
```
Perhaps this nested if statements could be refactored with guard clauses to make the happy path more apparent. Similarily with createEvent().
```suggestion
        String response = "";
        if (cmd.length < 2) {
            throw new DukeException("The description of a Deadline date cannot be empty.");
        }

        String[] deadlineDetails = cmd[1].split(" /by ", 2);

        if (deadlineDetails.length < 2) {
                throw new DukeException("Please indicate a time for the Deadline!");
        }

        try {
            Deadline deadline = new Deadline(deadlineDetails[0], deadlineDetails[1]);
            taskList.addTask(deadline);
            response = ui.showTaskAdded(deadline);
        } catch (DateTimeParseException e) {
            System.out.println(e.getMessage() + "\nPlease input deadline in yyyy-mm-dd");
        }
   
        return response;
```

Oddcorner commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,54 @@
+
+    abstract public String getTime();
+
+    public Boolean getIsDone() {return this.isDone;}
```
I think naming this function as "isDone" is more in line with the coding quality naming convention, what do you think?

Oddcorner commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,54 @@
+     * @throws DukeException
+     */
+    public static Task createTask(String[] description) throws DukeException {
+        switch (description[0]) {
+            case "T":
+                return new Todo(description[1], description[2]);
+            case "E":
+                return new Event(description[1], description[2], description[3]);
+            case "D":
+                return new Deadline(description[1], description[2], description[3]);
+            default:
+                throw new DukeException("Invalid Task");
+        }
```
switch and case statements should have the same indentation level to follow the coding standard guidelines.

Oddcorner reviewed (commented):
Overall code quality looks good! 👍 
The only places I found hard to follow/read is how the file is read/written in Stroage.java
Perhaps some method extractions will help with code quality there!

Good Job!

talkintomato committed:
changes made

talkintomato committed:
create HelpCmd

Help command allows user to see list of commands available for them to
use.

new users can learn the commands.

Lets's
* fix the UI of dialogBox

talkintomato committed:
change DialogBox

talkintomato committed:
add assertion to ensure userinput is not blank

talkintomato committed:
Merge pull request #2 from talkintomato/branch-A-Assertions

add assertion to ensure userinput is not blank

talkintomato committed:
Set theme jekyll-theme-slate

talkintomato committed:
edit help cmd technical error

talkintomato committed:
Merge pull request #3 from talkintomato/branch-C-Help

Branch c help

talkintomato committed:
minor edit to Readme

talkintomato committed:
fix FXML Launch

talkintomato committed:
fix FXML

talkintomato committed:
add description and expected outcome to Readme

talkintomato committed:
edit Readme format

talkintomato committed:
resolve merge conflicts with A-UserGuide

talkintomato committed:
add Ui.png

talkintomato committed:
edit dependencies for gradle.build to allow for windows os support

talkintomato committed:
clean up code for final production

talkintomato committed:
clean up files

talkintomato committed:
edit build.gradle for smoketest

talkintomato committed:
fixed bugs from ip quality feedback

talkintomato committed:
Merge branch 'master' of https://github.com/talkintomato/ip into master

talkintomato committed:
saved fixed

talkintomato committed:
save to home

talkintomato committed:
test merge

talkintomato committed:
solved javafx
